### DISCUSSION TITLE
Is it possible to use ANEs on desktop with ADL?

### DISCUSSION THREAD

**Discussion by Michoko92** [airsdk/Adobe-Runtime-Support#2008] [Q&A] [answered]
Hi,

I'd like to use the MediaPlayer ANE from Distriqt, and debug it while developping on desktop (PC). This ANE has a fallback system on PC, where it will use the default Flash sound system. However, I can't make it work.

Here is my ADL command line:
`adl.exe application-desktop.xml -profile extendedDesktop -extdir "D:\Projects\Libs\extensions"`

Here is an excerpt of my "application-desktop.xml" file:
```
<supportedProfiles>extendedDesktop</supportedProfiles>

<extensions>
	<extensionID>com.distriqt.Core</extensionID>
	<extensionID>com.distriqt.InAppBilling</extensionID>
	<extensionID>com.distriqt.Share</extensionID>
	<extensionID>com.distriqt.Application</extensionID>
	<extensionID>com.distriqt.ApplicationRater</extensionID>
	<extensionID>com.distriqt.Dialog</extensionID>
	 <extensionID>com.distriqt.MediaPlayer</extensionID>
</extensions>
```

But when I run the ADL command, I get the following error:
Error: Requested extension com.distriqt.Core could not be found.
The content cannot be loaded because there was a problem loading an extension: Error: Requested extension com.distriqt.Core could not be found.

I'm sure my "-extdir" path is good, as I use the same to package my executable build.

Is there a way to use ANEs on desktop in ADL, so we can debug our code easily? Or can we only make EXE builds with ADT (not sure of how to debug my code then?).

Thank you very much! :)

**Comment by ajwfrost** [comment, marked as answer]
You can use extensions with ADL but I believe it expects them to be extracted (as if they've been properly packaged/installed). So within your 'extensions' folder, it's going to expect a folder such as `com.distriqt.Core.ane` which contains files `mimetype`, `catalog.xml`, `library.swf` and the `META-INF\ANE` folder in which there should be the `extension.xml` and the platform folders.

i.e. unzip the ".ane" file into a folder of the same name...

**Reply by Michoko92** [reply]
Oh joy! It worked! :)

Thank you so much Andrew for your amazing support, as always! Cheers!

**Reply by marchbold** [reply]
@ajwfrost Is there anyway that we could improve this process? Definitely would be easier if we had similar options as adt i.e. with just providing the `extdir` ? It's not a major issue, just something we have found annoying at times :)

**Reply by ajwfrost** [reply]
@marchbold yes I was wondering that as I wrote the update.. we need to be able to load in the library (dll/dylib) via the operating system so it definitely needs to be extracted, but we could look at adding this step automatically i.e. if we can't find an ANE in the `extdir` via the existing method, we look at any .ane files that are there and see if any of these match it, if so we expand them and try again...

We can look at adding that as a bit of a convenience option, although it implies we'd need to be able to unzip the file (I've wondered for a while about including libzip or similar..)

**Reply by marchbold** [reply]
Interesting, I had assumed there must have been some form of zip utils in there already, considering you have to do something like that with the extensions in `adt` ? But yeah that would be the idea, just some sort of automatic unzipping (and if you are going to check for an existing one, make sure it's the same version :) had several issues with intellij in the past not updating correctly)

**Reply by ajwfrost** [reply]
`ADT` uses Java and the zip capabilities in there (`java.util.zip.*`) but it doesn't use the runtime; the AIR runtime doesn't have any zip-specific capabilities, but it does have the zlib and lzma compression and decompression capabilities (used for SWF files as well as in the `ByteArray` functions for compress/decompress). So actually looking at it, we can add an ActionScript package like `air.utils.zip` to expose basic zip functionality and use this to automatically extract the files we need from an ANE file.

Will put it into the backlog, for a later AIR version..

**Reply by 2jfw** [reply]
> You can use extensions with ADL but I believe it expects them to be extracted (as if they've been properly packaged/installed). So within your 'extensions' folder, it's going to expect a folder such as `com.distriqt.Core.ane` which contains files `mimetype`, `catalog.xml`, `library.swf` and the `META-INF\ANE` folder in which there should be the `extension.xml` and the platform folders.
> 
> i.e. unzip the ".ane" file into a folder of the same name...

There is no need to manually extract the ANE's - *FlashBuilder* and also *IntellJ* will do this for you - just did a quick test with the `DesktopToastANE` (https://github.com/tuarua/DesktopToastANE) and it is working fine. The ANE only needs to be linked in the same way as you would do with any SWC

**Reply by marchbold** [reply]
@2jfw Yeah the IDEs do this for you, but if you are trying to automate builds and tests etc with applications that use extensions you'll have to do this manually in your scripts.

**Reply by 2jfw** [reply]
> @2jfw Yeah the IDEs do this for you, but if you are trying to automate builds and tests etc with applications that use extensions you'll have to do this manually in your scripts.

Ah, I see - thanks for clarifying, Michael!
